tests/bulk_create/tests.py

```diff
@@ -217,3 +218,11 @@ def test_set_pk_and_query_efficiency(self):
         self.assertEqual(Country.objects.get(pk=countries[1].pk), countries[1])
         self.assertEqual(Country.objects.get(pk=countries[2].pk), countries[2])
         self.assertEqual(Country.objects.get(pk=countries[3].pk), countries[3])
+
+    @skipUnlessDBFeature('can_return_ids_from_bulk_insert')
+    def test_set_pk_and_allow_creating_related(self):
+        union_eu = Union.objects.create(name='EU')
+        country_nl = Country(name='Netherlands', iso_two_letter='NL')
+        with self.assertNumQueries(1):
+            Country.objects.bulk_create([country_nl])
+        union_eu.countries.add(country_nl)
```

Comment:
As the underlying issue is state initialization and many-to-many assignment failure is only a symptom I suggest testing for `country_nl._state.adding` and `country_nl._state.db` instead. This should avoid defining the `Union` model at the same time.